mp4parse/src/lib.rs

```diff
@@ -122,70 +120,310 @@ impl<'a, T: Read> Read for OffsetReader<'a, T> {
-/// `src` accordingly.
-///
-/// With the `mp4parse_fallible` feature disabled, this is wrapper around
-/// `std::io::Read::read_to_end()`.
-fn read_to_end<T: Read>(src: &mut Take<T>, buf: &mut Vec<u8>) -> std::result::Result<usize, ()> {
-    #[cfg(feature = "mp4parse_fallible")]
-    {
-        let limit: usize = src.limit().try_into().map_err(|_| ())?;
-        FallibleVec::try_reserve(buf, limit)?;
-        let bytes_read = src.read_to_end(buf).map_err(|_| ())?;
-        Ok(bytes_read)
+    /// TryVec is a thin wrapper around std::vec::Vec to provide support for
+    /// fallible allocation. Though the fallible allocator is only enabled
+    /// with the mp4parse_fallible feature, the API differences from stdlib
+    /// Vec ensure that all operations which allocate return a Result. For the
+    /// most part, this simply means adding a Result return value to Vec
+    /// functions which return nothing or a non-Result value. However, Vec
+    /// implements some traits whose API cannot communicate failure, but which
+    /// do require allocation, so it is important that TryVec does not
+    /// implement these traits.
+    ///
+    /// Specifically, do not implement any of the following traits:
+    /// - Clone
+    /// - Extend
+    /// - From
+    /// - FromIterator
+    ///
+    /// This list may not be exhaustive. Exercise caution when implementing
+    /// any new traits to ensure they won't potentially allocate in a way that
+    /// can't return a Result to indicate allocation failure.
+    impl<T> TryVec<T> {
+        pub fn new() -> Self {
+            Self { inner: Vec::new() }
+        }
+
+        pub fn with_capacity(capacity: usize) -> Result<Self> {
+            let mut v = Self::new();
+            v.reserve(capacity)?;
+            Ok(v)
+        }
+
+        pub fn append(&mut self, other: &mut Self) -> Result<()> {
+            self.reserve(other.inner.len())?;
+            self.inner.append(&mut other.inner);
+            Ok(())
+        }
+
+        pub fn as_mut_slice(&mut self) -> &mut [T] {
+            self
+        }
+
+        pub fn as_slice(&self) -> &[T] {
+            self
+        }
+
+        pub fn clear(&mut self) {
+            self.inner.clear()
+        }
+
+        #[cfg(test)]
+        pub fn into_inner(self) -> Vec<T> {
+            self.inner
+        }
+
+        pub fn is_empty(&self) -> bool {
+            self.inner.is_empty()
+        }
+
+        pub fn iter_mut(&mut self) -> IterMut<T> {
+            IterMut {
+                inner: self.inner.iter_mut(),
+            }
+        }
+
+        pub fn iter(&self) -> Iter<T> {
+            Iter {
+                inner: self.inner.iter(),
+            }
+        }
+
+        pub fn pop(&mut self) -> Option<T> {
+            self.inner.pop()
+        }
+
+        pub fn push(&mut self, value: T) -> Result<()> {
+            self.reserve(if self.inner.capacity() == 0 { 4 } else { 1 })?;
```

Comment:
After pulling in this little bit of logic, the only remaining dependency on `FallibleVec` is `try_reserve`. I wonder if we should consider absorbing that into this repo.

Reply:
> I wonder if we should consider absorbing that into this repo.

:+1:

Reply:
Will follow up on this with my next PR to handle the other potentially allocating types.